hpct-ldap-server/src/charm.py

```diff
@@ -0,0 +1,137 @@
+
+    def _on_add_group_action(self, event):
+        """Handle add-group action."""
+        if not self.unit.is_leader():
+            event.fail("The action can be run only on leader unit.")
+            return
+        replicas = self.model.get_relation("replicas")
+        domain = replicas.data[self.app].get("domain")
+        if not domain:
+            event.fail("Domain has not been set.")
+            return
+        self.ldapserver_manager.add_group(
+            event.params["admin-passwd"],
+            domain,
+            event.params["gid"],
+            event.params["group"],
+        )
+
+    def _on_add_user_action(self, event):
+        """Handle add-user action."""
+        if not self.unit.is_leader():
+            event.fail("The action can be run only on leader unit.")
+            return
+        replicas = self.model.get_relation("replicas")
+        domain = replicas.data[self.app].get("domain")
+        self.ldapserver_manager.add_user(
+            event.params["admin-passwd"],
+            domain,
+            event.params["gecos"],
+            event.params["gid"],
+            event.params["homedir"],
+            event.params["shell"],
+            event.params["passwd"],
+            event.params["uid"],
+            event.params["user"],
+        )
+
+    def _on_install(self, event):
+        """Handle install event."""
+        logger.info("Install")
+        self.ldapserver_manager.install()
+        if not self.ldapserver_manager.is_installed():
+            logger.error("Install failed, required packages not found.")
+
+    def _on_configure_action(self, event):
+        """Handle set-config action."""
+        if not self.unit.is_leader():
+            event.fail("The action can be run only on leader unit.")
+            return
+        self.ldapserver_manager.configure(
+            event.params["admin-passwd"],
+            event.params["domain"],
+            event.params["org"],
+        )
+        self.ldapserver_manager.tls_gen(event.params["org"])
+        replicas = self.model.get_relation("replicas")
+        replicas.data[self.app].update(
+            {
+                "domain": event.params["domain"],
+                "org": event.params["org"],
+            }
+        )
+
+    def _on_start(self, event):
+        """Handle start event."""
+        self.ldapserver_manager.start()
+        self.unit.status = ActiveStatus("LDAP Server Started")
+
+    def _on_ldap_auth_relation_changed(self, event):
+        """Handle the ldap-auth action."""
+        # Get ldap-auth relation
+        auth_relation = self.model.get_relation("ldap-auth")
+        # Get Replicas relation
+        replicas = self.model.get_relation("replicas")
+        domain = replicas.data[self.app].get("domain")
+        if not domain:
+            logger.info(f"domain: {domain}")
+            self.unit.status = WaitingStatus("Waiting for set-config action to be run")
+            event.defer()
```

Comment:
Are you wanting an event to be triggered indefinitely until the domain is set via the action.

Reply:
No, will remove the event.defer